### PR TITLE
Python 2/3 first steps

### PR DESCRIPTION
**Pull request by Cabalist**
Hey there,

I really like your project!  I spent a little time migrating some things over to ease a transition to Python3 while still maintaining Python 2 support.  From what I can tell in the existing code the minimum requirement was Python 2.5.  The changes I have made increase that to 2.6 and that is mostly due to how Exceptions are printed.  `Exception().message` was introduced in 2.5 and depreciated in 2.6 and removed in 3.0.  

This is not complete Python3 support.  There are still some places where Bytes and Strings are being inconsistently handled.  Without some tests I wasn't confident to go in and change those things.  If you have interest I think I can bang out some that would work.  

The other noteworthy thing I replaced `.iteritems()` and `xrange()` with `.items()` and `range()` respectively.  I *believe* in the use cases that these very minimally affect memory usage in Py2 and shouldn't introduce any performance issues.  The things that are being iterated over are relatively small.  Correct me if I am wrong! 😃

### CODE REVIEW

**Comment by Cabalist** [comment]
OH!  And there were places were you had a mutable default argument (a dict of `callbacks`) this was modified to prevent bugs.

File: ptftplib/notify.py [hunk 1]
```diff
@@ -58,43 +59,43 @@ class StreamEngine(logging.StreamHandler):
     """Simple stream log handler, similar to what you get using
     logging.basicConfig."""
 
-    def __init__(self, stream, loglevel, format):
+    def __init__(self, stream, loglevel, format_):
```
Comment (mpetazzoni) [inline ptftplib/notify.py]: Why this change?

File: ptftplib/notify.py [hunk 2]
```diff
@@ -142,14 +143,16 @@ def emit(self, record):
         """Call the defined callback for the transfer state found in the log
         record."""
 
-        callable = self.callbacks.get(record.state, self._nop)
-        callable(host=record.host,
-                 port=record.port,
-                 file=record.file,
-                 state=record.state)
+        callable_ = self.callbacks.get(record.state, self._nop)
```
Comment (mpetazzoni) [inline ptftplib/notify.py]: Why this change too?

File: ptftplib/tftpclient.py [hunk 3]
```diff
@@ -19,6 +20,14 @@
 # along with pTFTPd.  If not, see <http://www.gnu.org/licenses/>.
 
 
+from __future__ import print_function
+
+try:
+    # noinspection PyShadowingBuiltins
+    input = raw_input  # Py2
```
Comment (mpetazzoni) [inline ptftplib/tftpclient.py]: Should we consider just using `six`?

```python
from six import input
```

File: ptftplib/tftpserver.py [hunk 4]
```diff
@@ -46,6 +46,11 @@
 from . import proto
 from . import state
 
+try:
+    import SocketServer as socketserver  # Py2
```
Comment (mpetazzoni) [inline ptftplib/tftpserver.py]: Same here, we can use `six`:

```python
from six import socketserver
```

File: ptftplib/tftpserver.py [hunk 5]
```diff
@@ -72,7 +77,8 @@ class TFTPServerConfigurationError(Exception):
     """The configuration of the pTFTPd is incorrect."""
 
 
-class TFTPServerHandler(SocketServer.DatagramRequestHandler):
+# noinspection PyPep8Naming
```
Comment (mpetazzoni) [inline ptftplib/tftpserver.py]: Why do you need this override?

File: ptftplib/notify.py (hunk 1 above)
Comment (Cabalist) [inline ptftplib/notify.py]: `format` is a built-in keyword.  Overwriting it can cause all sorts of issues.  I added the underscore at the end but maybe a better name could be `fmt`?

File: ptftplib/notify.py (hunk 2 above)
Comment (Cabalist) [inline ptftplib/notify.py]: same as above.  `callable` is a built-in

File: ptftplib/tftpclient.py (hunk 3 above)
Comment (Cabalist) [inline ptftplib/tftpclient.py]: I like `six` for big libraries but since the only really compatibility thing we are looking at is `input` vs `raw_input` it is easy to just overwrite this one. You avoid a dependency as well.  :)

File: ptftplib/tftpserver.py (hunk 5 above)
Comment (Cabalist) [inline ptftplib/tftpserver.py]: PEP8 doesn't want your methods to be in Camelcase.  It wants `TFTPServerHandler.serveRRQ` to be `TFTPServerHandler.server_rrq`  I added that to make my linter stop whining because changing those method names would be API breakage.

File: ptftplib/notify.py (hunk 2 above)
Comment (Cabalist) [inline ptftplib/notify.py]: https://docs.python.org/3/library/functions.html#callable

File: ptftplib/notify.py (hunk 1 above)
Comment (Cabalist) [inline ptftplib/notify.py]: https://docs.python.org/3/library/functions.html#format

Comment (mpetazzoni) [inline ptftplib/notify.py]: Ah, thanks; I wasn't sure if it was one. I guess I'd rather use `fmt` as the variable name then.

File: ptftplib/notify.py (hunk 2 above)
Comment (mpetazzoni) [inline ptftplib/notify.py]: I guess it is now... Name this one `callback` then.

File: ptftplib/tftpclient.py (hunk 3 above)
Comment (mpetazzoni) [inline ptftplib/tftpclient.py]: Makes sense, I guess I can live with that.

**Comment by Cabalist** [comment]
Variables have been renamed. :)

**Comment by mpetazzoni** [comment]
Awesome, thanks for appeasing my attention to detail.
